### ISSUE TITLE
Predicting adjacency matrix from nodes features (supervised)

### ISSUE THREAD

**Issue by BRO-HAMMER** [danielegrattarola/spektral#69]
Hello, I´ve just discovered this library and I´m wondering what architecture would be useful for the following task, in case this framework is well suited for this:

I have examples of "batches" of data points that form clusters (some points don't belong to any cluster). Treating them as nodes, I can extract the adjacency matrices of the disjoint graphs, and the nodes features for training, in order to predict the adj matrix (links) of new batches from their node features. In other words, my objective is to train a net to cluster the data points in a supervised way.

I was appraching this with RNNs and FC nets, using a bunch of sigmoid activations as the output layer and binary-crossentropy to approximate the adj matrix, but after finding out about GNNs, they sounded more appropriate for a task of this nature. If someone has any idea, I'd appreciate it.

Thanks!

**Comment by danielegrattarola** [comment]
GNNs are useful when you have a given graph as input. In your case, if I understand correctly, it seems like you want to generate a graph from simple vector data, although I'm not sure I understand. Why do you want to predict links in order to cluster? Isn't it better to frame it as a classification problem?

There are some papers where they learn predict a graph from the nodes also using GNN (notably https://arxiv.org/abs/1802.04687 comes to mind), but it really depends on the particular task and I'm not sure it's what you're trying to do.

**Comment by BRO-HAMMER** [comment]
First of all, thank you very much for your reply and the reference to the paper. You are correct, I have data that can represented as vectors (combining categorical and continuous features) and also the ground-truth for how those data points should be clustered.

I´ll be more specific about the task: for a given newspaper page, I'm using object detection to extract the location and category of every component (headlines, bodies of text, images, captions, etc.). I´m trying to figure out if it's possible to group those elements in distinct articles from the visual information only (without delving into NLP).

The problem is that (at least for this initial representation) traditional unsupervised clustering algorithms woldn´t work: points in the same cluster are not necessarily close to each other in vector space). Both the number of clusters (articles) and the number of elements vary in each page. I´d like to create an architecture that could learn the patterns of association to predict the links between vectors.

My current approach is treating each page as a sequence of vectors , feeding it to a Bidirectional RNN (many to many) and then predicting each bit of the adjacency matrix as a binary classification problem.

Based on the paper you pointed out, I was wondering if instead of the RNN, It would make sense to use a GNN starting from a fully-connected graph to learn a better representation that I could then use to predict the links that should be kept/dropped.

Thanks again, I don´t want to take more of your time, if you have any idea that I could try within this framework great, if not you can close the issue, no problem.

**Comment by danielegrattarola** [comment]
It sounds like a nice project! 

The first thing I would do is to try a fully connected graph between the different objects and using a couple of message-passing layers to propagate information on your graph. I would suggest to use GAT so that the GNN can learn which neighbours are relevant for each node. 
In this way you will have that each node will have incorporated information from all over the page in its representation. 

Then, you can try a simple inner product decoder to compute the similarity between each pair of nodes (see `spektral.layers.InnerProduct`) and train it with the real adjacency matrix as supervision.

I'm not sure if this is exactly what you were looking for, but if you want to chat more about your project shoot me an email.
